.config/doom/config.org
#+title: My personal emacs configuration.
#+PROPERTY: header-args:emacs-lisp :tangle ./config.el

* Babel tangle
This makes sure that everytime this org file gets saved the config.el file is updated.
#+begin_src emacs-lisp :tangle yes
(defun mp/tangle-on-save-org-mode-file()
  (when (string= (message "%s" major-mode) "org-mode")
    (org-babel-tangle)))

(add-hook 'after-save-hook 'mp/tangle-on-save-org-mode-file)
#+end_src
* Theme
We use the package circadian to change between a light and dark theme.
#+begin_src emacs-lisp :tangle yes
(setq modus-themes-headings
      '((1 . (variable-pitch 1.5)) ;; 1 means first heading level (document title)
        (2 . (1.3)) ;;
        (agenda-date . (1.3))
        (agenda-structure . (variable-pitch light 1.8))
        (t . (1.1))))

(load-theme 'modus-vivendi-tinted t)
#+end_src
** Toggle light dark theme
#+begin_src emacs-lisp
(map! :leader
      :desc "Light and dark theme"
      "t t" 'modus-themes-toggle)
#+end_src
* UI
** Line numbers
#+begin_src emacs-lisp :tangle yes
(column-number-mode)
(global-display-line-numbers-mode 0)
(setq display-line-numbers-type 'relative) ;; This sets relative line numbers.
#+end_src
** Font
#+begin_src emacs-lisp :tangle yes
(setq
    doom-font (font-spec :family "JetBrainsMono Nerd Font Mono" :size 16 :weight 'regular) doom-big-font (font-spec :family "JetBrainsMono Nerd Font Mono" :size 24)
    doom-variable-pitch-font (font-spec :family "SF Pro Text")
     )
#+end_src
** Modeline
#+begin_src emacs-lisp
(setq doom-modeline-bar-height 25)
(setq doom-modeline-time-icon t)
(setq doom-modeline-project-name t)
#+end_src
* Keybindings
#+begin_src emacs-lisp :tangle yes
(use-package! evil-escape
    :init
    (setq evil-escape-key-sequence "jk"))
#+end_src
Open Weekly agenda
#+begin_src emacs-lisp :tangle yes
(map! :leader
      :desc "Weekly agenda"
      "o w" #'org-agenda-list)
#+end_src
Searching
** General documents
 #+begin_src emacs-lisp :tangle yes
(defun my/search-documents ()
  "Search files in ~/Documents using consult fd/find."
  (interactive)
  (let ((default-directory "~/Documents/"))
    (consult-fd)))

(map! :leader
      :desc "Find file in Documents"
      "d f" #'my/search-documents)
 #+end_src
** Papers
 #+begin_src emacs-lisp :tangle yes
(defun my/search-papers ()
  "Search files in ~/Documents using consult fd/find."
  (interactive)
  (let ((default-directory "~/Documents/Papers"))
    (consult-fd)))

(map! :leader
      :desc "Find paper"
      "d p" #'my/search-papers)
 #+end_src
** Citations
#+begin_src emacs-lisp :tangle yes
(map! :leader
      :desc "Find citation"
      "d c" #'citar-open-entry)
#+end_src
* Org
** Org Roam
We want to use this package as our note taking system.
*** Directory
Org roam directory will be located in the Zettelkasten directory.
#+begin_src emacs-lisp :tangle yes
(setq org-roam-directory "~/Zettelkasten")
#+end_src
*** Turn of completions
#+begin_src emacs-lisp :tangle yes
(setq org-roam-completion-everywhere nil)
#+end_src
** Agenda
#+begin_src emacs-lisp :tangle yes
(setq org-journal-enable-agenda-integration t
      org-icalendar-store-UID t
      org-icalendar-include-todo "all"
      org-icalendar-combined-agenda-file "~/Org/emacs.ics")
#+end_src
** Alert
#+begin_src emacs-lisp :tangle yes
(use-package org-alert
  :ensure t)
(setq alert-default-style 'libnotify)
#+end_src
** UI
#+begin_src emacs-lisp
;; (setq org-ellipsis "[..]") ;; symbol if header is closed
;; (set-face-attribute 'org-ellipsis nil :inherit 'default :box nil)
#+end_src
** Olivetti
#+begin_src emacs-lisp :tangle yes
;; (use-package olivetti
;;   :ensure t
;;   :hook (org-mode . olivetti-mode)
;;   :config
;;   (setq olivetti-body-width 200))
#+end_src
** Org download
#+begin_src emacs-lisp :tangle yes
(require 'org-download)

;; Drag-and-drop to `dired`
(add-hook 'dired-mode-hook 'org-download-enable)
#+end_src
* Email
#+begin_src emacs-lisp :tangle yes
;; (require 'mu4e)
(use-package mu4e
        :ensure nil
        :config
        (setq mu4e-change-filenames-when-moving t)
        (setq mu4e-update-interval (* 10 60))
        (setq mu4e-get-mail-command "mbsync -a")
        (setq mu4e-maildir (expand-file-name "~/Mail"))
        (setq mu4e-headers-show-threads nil)
        (setq mu4e-headers-include-related nil)

        ;; Make sure plain text mails flow correctly for recipients
        (setq
            mail-user-agent 'mu4e-user-agent
            sendmail-program (executable-find "msmtp")
            send-mail-function 'message-send-mail-with-sendmail
            message-send-mail-function 'message-send-mail-with-sendmail

            message-kill-buffer-on-exit t
            message-sendmail-f-is-evil t
            message-sendmail-extra-arguments '("--read-envelope-from")
            message-sendmail-envelope-from 'header)
         )
        ;; (setq mu4e-compose-format-flowed t)
        ;; (setq sendmail-program (executable-find "msmtp"))
        ;; ;; Configure the function to use for sending mail
        ;; ;; (setq message-send-mail-function 'smtpmail-send-it)
        ;; (setq send-mail-function 'sendmail-send-it
        ;;     message-send-mail-function 'sendmail-send-it)
        ;; )
(setq mu4e-contexts
      (list
       ;; Work account
       (make-mu4e-context
        :name "FAU"
        :enter-func (lambda () (mu4e-message "Switched to Work context"))
        :match-func (lambda (msg)
                (when msg
                        (string-prefix-p "/FAU" (mu4e-message-field msg :maildir))))
        :vars '((user-mail-address . "markus.pirke@fau.de")
                (user-full-name    . "Markus Pirke")
                (mu4e-sent-folder  . "/FAU/Sent")
                (mu4e-drafts-folder . "/FAU/Drafts")
                (mu4e-trash-folder  . "/FAU/Trash")
                (mu4e-refile-folder . "/FAU/Archive")))

       (make-mu4e-context
        :name "iCloud"
        :enter-func (lambda () (mu4e-message "Switched to Personal context"))
        :match-func (lambda (msg)
                (when msg
                        (string-prefix-p "/iCloud" (mu4e-message-field msg :maildir))))
        :vars '((user-mail-address . "markus.pirke@icloud.com")
                (user-full-name    . "Markus Pirke")
                (mu4e-sent-folder  . "/iCloud/Sent Messages")
                (mu4e-drafts-folder . "/iCloud/Drafts")
                (mu4e-trash-folder  . "/iCloud/Deleted Messages")
                (mu4e-refile-folder . "/iCloud/Archive")))))

(add-to-list 'mu4e-bookmarks
  ;; bookmark for message that require quick attention
  '( :name "Unread Work"
     :key  ?w
     :query "flag:unread AND NOT flag:trashed AND maildir:/FAU/INBOX"))

(add-to-list 'mu4e-bookmarks
  ;; bookmark for message that require quick attention
  '( :name "Unread Personal"
     :key  ?p
     :query "flag:unread AND NOT flag:trashed AND maildir:/iCloud/INBOX"))
#+end_src
* Utils
** Multiple languages within Emacs
We can change within emacs between multiple languages without changing the system wide keyboard layout!
#+begin_src emacs-lisp :tangle yes
(setq default-input-method "german")
(setq default-transient-input-method "german")
#+end_src
* GPTEL
#+begin_src emacs-lisp :tangle yes
(gptel-make-ollama "Ollama"             ;Any name of your choosing
  :host "localhost:11434"               ;Where it's running
  :stream t                             ;Stream responses
  :models '(ministral-3:8b
            ministral-3:3b
            qwen3:4b
            qwen3-vl:4b
            lfm2.5-thinking:latest))
#+end_src
* Citations
#+begin_src emacs-lisp :tangle yes
(use-package citar
  :custom
  (citar-bibliography '("~/Documents/references.bib")))

(defun my/citar-insert-latex-cite ()
  "Insert a \\cite{key} at point using citar."
  (interactive)
  (let ((key (car (citar-select-refs))))
    (insert (format "\\cite{%s}" key))))

(map! :leader
      :desc "Insert citation" "i c" #'my/citar-insert-latex-cite)
#+end_src

* Jinx
This seems to be a great package for finding spelling mistakes. Some conflicts with flyspell-mode.
One needs to install a two external dependencies.
#+begin_src bash
homebrew install enchant
homebrew install pkgconf
#+end_src
The latter one had conflicts on my system, as I already had installed something similar called /pkg-config/. But it worked nonetheless.
#+begin_src emacs-lisp :tangle yes
(use-package jinx
  :ensure t
  :hook ((LaTeX-mode . jinx-mode)
         (latex-mode . jinx-mode)
         (org-mode . jinx-mode)
         (text-mode . jinx-mode))
  :config (setq jinx-languages "en_US,de")
  )
;; this turns of the flyspell-mode when an org document is opened
;; disable flyspell-mode
(setq-default spell-checking-enable-by-default nil)
(remove-hook 'org-mode-hook #'flyspell-mode)
(remove-hook 'latex-mode-hook #'flyspell-mode)
(remove-hook 'LaTeX-mode-hook #'flyspell-mode)
(remove-hook 'text-mode-hook #'flyspell-mode)
;; Shortcut for correct word
(map! :leader
      (:prefix ("e" . "edit")
      :desc "Correct word"
      "w" #'jinx-correct-word))
#+end_src


.config/doom/config.el
(defun mp/tangle-on-save-org-mode-file()
  (when (string= (message "%s" major-mode) "org-mode")
    (org-babel-tangle)))

(add-hook 'after-save-hook 'mp/tangle-on-save-org-mode-file)

(setq modus-themes-headings
      '((1 . (variable-pitch 1.5)) ;; 1 means first heading level (document title)
        (2 . (1.3)) ;;
        (agenda-date . (1.3))
        (agenda-structure . (variable-pitch light 1.8))
        (t . (1.1))))

(load-theme 'modus-vivendi-tinted t)

(map! :leader
      :desc "Light and dark theme"
      "t t" 'modus-themes-toggle)

(column-number-mode)
(global-display-line-numbers-mode 0)
(setq display-line-numbers-type 'relative) ;; This sets relative line numbers.

(setq
    doom-font (font-spec :family "JetBrainsMono Nerd Font Mono" :size 16 :weight 'regular) doom-big-font (font-spec :family "JetBrainsMono Nerd Font Mono" :size 24)
    doom-variable-pitch-font (font-spec :family "SF Pro Text")
     )

(setq doom-modeline-bar-height 25)
(setq doom-modeline-time-icon t)
(setq doom-modeline-project-name t)

(use-package! evil-escape
    :init
    (setq evil-escape-key-sequence "jk"))

(map! :leader
      :desc "Weekly agenda"
      "o w" #'org-agenda-list)

(defun my/search-documents ()
  "Search files in ~/Documents using consult fd/find."
  (interactive)
  (let ((default-directory "~/Documents/"))
    (consult-fd)))

(map! :leader
      :desc "Find file in Documents"
      "d f" #'my/search-documents)

(defun my/search-papers ()
  "Search files in ~/Documents using consult fd/find."
  (interactive)
  (let ((default-directory "~/Documents/Papers"))
    (consult-fd)))

(map! :leader
      :desc "Find paper"
      "d p" #'my/search-papers)

(map! :leader
      :desc "Find citation"
      "d c" #'citar-open-entry)

(setq org-roam-directory "~/Zettelkasten")

(setq org-roam-completion-everywhere nil)

(setq org-journal-enable-agenda-integration t
      org-icalendar-store-UID t
      org-icalendar-include-todo "all"
      org-icalendar-combined-agenda-file "~/Org/emacs.ics")

(use-package org-alert
  :ensure t)
(setq alert-default-style 'libnotify)

;; (setq org-ellipsis "[..]") ;; symbol if header is closed
;; (set-face-attribute 'org-ellipsis nil :inherit 'default :box nil)

;; (use-package olivetti
;;   :ensure t
;;   :hook (org-mode . olivetti-mode)
;;   :config
;;   (setq olivetti-body-width 200))

(require 'org-download)

;; Drag-and-drop to `dired`
(add-hook 'dired-mode-hook 'org-download-enable)

;; (require 'mu4e)
(use-package mu4e
        :ensure nil
        :config
        (setq mu4e-change-filenames-when-moving t)
        (setq mu4e-update-interval (* 10 60))
        (setq mu4e-get-mail-command "mbsync -a")
        (setq mu4e-maildir (expand-file-name "~/Mail"))
        (setq mu4e-headers-show-threads nil)
        (setq mu4e-headers-include-related nil)

        ;; Make sure plain text mails flow correctly for recipients
        (setq
            mail-user-agent 'mu4e-user-agent
            sendmail-program (executable-find "msmtp")
            send-mail-function 'message-send-mail-with-sendmail
            message-send-mail-function 'message-send-mail-with-sendmail

            message-kill-buffer-on-exit t
            message-sendmail-f-is-evil t
            message-sendmail-extra-arguments '("--read-envelope-from")
            message-sendmail-envelope-from 'header)
         )
        ;; (setq mu4e-compose-format-flowed t)
        ;; (setq sendmail-program (executable-find "msmtp"))
        ;; ;; Configure the function to use for sending mail
        ;; ;; (setq message-send-mail-function 'smtpmail-send-it)
        ;; (setq send-mail-function 'sendmail-send-it
        ;;     message-send-mail-function 'sendmail-send-it)
        ;; )
(setq mu4e-contexts
      (list
       ;; Work account
       (make-mu4e-context
        :name "FAU"
        :enter-func (lambda () (mu4e-message "Switched to Work context"))
        :match-func (lambda (msg)
                (when msg
                        (string-prefix-p "/FAU" (mu4e-message-field msg :maildir))))
        :vars '((user-mail-address . "markus.pirke@fau.de")
                (user-full-name    . "Markus Pirke")
                (mu4e-sent-folder  . "/FAU/Sent")
                (mu4e-drafts-folder . "/FAU/Drafts")
                (mu4e-trash-folder  . "/FAU/Trash")
                (mu4e-refile-folder . "/FAU/Archive")))

       (make-mu4e-context
        :name "iCloud"
        :enter-func (lambda () (mu4e-message "Switched to Personal context"))
        :match-func (lambda (msg)
                (when msg
                        (string-prefix-p "/iCloud" (mu4e-message-field msg :maildir))))
        :vars '((user-mail-address . "markus.pirke@icloud.com")
                (user-full-name    . "Markus Pirke")
                (mu4e-sent-folder  . "/iCloud/Sent Messages")
                (mu4e-drafts-folder . "/iCloud/Drafts")
                (mu4e-trash-folder  . "/iCloud/Deleted Messages")
                (mu4e-refile-folder . "/iCloud/Archive")))))

(add-to-list 'mu4e-bookmarks
  ;; bookmark for message that require quick attention
  '( :name "Unread Work"
     :key  ?w
     :query "flag:unread AND NOT flag:trashed AND maildir:/FAU/INBOX"))

(add-to-list 'mu4e-bookmarks
  ;; bookmark for message that require quick attention
  '( :name "Unread Personal"
     :key  ?p
     :query "flag:unread AND NOT flag:trashed AND maildir:/iCloud/INBOX"))

(setq default-input-method "german")
(setq default-transient-input-method "german")

(gptel-make-ollama "Ollama"             ;Any name of your choosing
  :host "localhost:11434"               ;Where it's running
  :stream t                             ;Stream responses
  :models '(ministral-3:8b
            ministral-3:3b
            qwen3:4b
            qwen3-vl:4b
            lfm2.5-thinking:latest))

(use-package citar
  :custom
  (citar-bibliography '("~/Documents/references.bib")))

(defun my/citar-insert-latex-cite ()
  "Insert a \\cite{key} at point using citar."
  (interactive)
  (let ((key (car (citar-select-refs))))
    (insert (format "\\cite{%s}" key))))

(map! :leader
      :desc "Insert citation" "i c" #'my/citar-insert-latex-cite)

(use-package jinx
  :ensure t
  :hook ((LaTeX-mode . jinx-mode)
         (latex-mode . jinx-mode)
         (org-mode . jinx-mode)
         (text-mode . jinx-mode))
  :config (setq jinx-languages "en_US,de")
  )
;; this turns of the flyspell-mode when an org document is opened
;; disable flyspell-mode
(setq-default spell-checking-enable-by-default nil)
(remove-hook 'org-mode-hook #'flyspell-mode)
(remove-hook 'latex-mode-hook #'flyspell-mode)
(remove-hook 'LaTeX-mode-hook #'flyspell-mode)
(remove-hook 'text-mode-hook #'flyspell-mode)
;; Shortcut for correct word
(map! :leader
      (:prefix ("e" . "edit")
      :desc "Correct word"
      "w" #'jinx-correct-word))
